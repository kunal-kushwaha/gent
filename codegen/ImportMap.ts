import path from 'path';
import _ from 'lodash';
import { CodeBuilder } from '@elg/tscodegen';

export type ImportMap = { [modulePath: string]: string[] | undefined };

function mergeImportMaps(maps: ImportMap[]): ImportMap {
  function merger(objValue: string[] | undefined, srcValue: string[] | undefined) {
    return [...(objValue ?? []), ...(srcValue ?? [])];
  }

  return maps.reduce((previousValue, currentValue) =>
    _.mergeWith(previousValue, currentValue, merger),
  );
}

function getDefaultNameFromModulePath(modulePath: string): string {
  const rawModuleName = path.basename(modulePath);
  let importableName = rawModuleName;

  // Handle snake/kebab case
  importableName = _.camelCase(rawModuleName);

  // If first char in original name is uppercase (e.g. class, enum names),
  // retain that uppercase first char.
  if (rawModuleName[0] === rawModuleName[0].toUpperCase()) {
    importableName = _.upperFirst(importableName);
  }

  return importableName;
}

export function buildImportLines(importMaps: ImportMap[], codeBuilder: CodeBuilder): CodeBuilder {
  const importMap = mergeImportMaps(importMaps);
  Object.entries(importMap).forEach(([modulePath, imports]) => {
    const importStatementComponents = [];

    if (imports?.includes('default')) {
      importStatementComponents.push(getDefaultNameFromModulePath(modulePath));
    }

    const otherImports = _.without(_.uniq(imports), 'default').sort();
    if (otherImports.length > 0) {
      importStatementComponents.push(`{ ${otherImports.join(', ')} }`);
    }

    if (importStatementComponents.length === 0) {
      codeBuilder.addLine(`import '${modulePath}';`);
    } else {
      codeBuilder.addLine(`import ${importStatementComponents.join(', ')} from '${modulePath}';`);
    }
  });
  return codeBuilder;
}
