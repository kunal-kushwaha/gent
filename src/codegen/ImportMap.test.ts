import { buildImportLines } from "./ImportMap";
import { CodeBuilder } from "@elg/tscodegen";

describe(buildImportLines, () => {
  test("should ignore empty map", () => {
    const builtCode = buildImportLines([{}], new CodeBuilder({})).toString();
    expect(builtCode).toBe("");
  });

  test("should add lines for each module", () => {
    const maps = [
      {
        a: ["one"],
        b: ["two"],
      },
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });

  test("should handle default-only module import", () => {
    const maps = [
      {
        camelCaseModule: ["default"],
        PascalCaseModule: ["default"],
        "../../pathModule": ["default"],
        "kebab-case-module": ["default"],
        snake_case_module: ["default"],
        "weird_mixed-separatorModule": ["default"],
        "dot.separated.module": ["default"],
        "@elg/scoped-package": ["default"],

        // Produces an invalid import. Intentional since there's no name we can
        // extract.
        ".": ["default"],
      },
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });

  test("should handle destructure-only module import", () => {
    const maps = [
      {
        a: ["one", "two"],
        b: ["two"],
      },
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });

  test("should handle default+destructure module import", () => {
    const maps = [
      {
        a: ["one", "default", "two"],
        b: ["default", "three"],
      },
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });

  test("should handle empty module import", () => {
    const maps = [
      {
        a: [],
        b: undefined,
      },
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });

  test("should merge and deduplicate multiple import maps", () => {
    const maps = [
      {
        a: ["default"],
        b: ["default"],
      },
      // Duplicate map
      {
        a: ["default"],
        b: ["default"],
      },
      {
        a: ["default"], // Duplicated
        b: ["two"], // Same module, different import
        // More modules
        c: ["zero"],
        "./../../moreCharacters": ["minusOne"],
      },
      {
        c: ["one"],
      },
      {
        d: [],
        e: undefined,
      },
      // Empty map
      {},
    ];
    const builtCode = buildImportLines(maps, new CodeBuilder({})).toString();
    expect(builtCode).toMatchSnapshot();
  });
});
