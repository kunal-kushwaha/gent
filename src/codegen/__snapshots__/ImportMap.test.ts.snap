// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildImportLines should add lines for each module 1`] = `
"import { one } from 'a';
import { two } from 'b';
"
`;

exports[`buildImportLines should handle default+destructure module import 1`] = `
"import a, { one, two } from 'a';
import b, { three } from 'b';
"
`;

exports[`buildImportLines should handle default-only module import 1`] = `
"import camelCaseModule from 'camelCaseModule';
import PascalCaseModule from 'PascalCaseModule';
import pathModule from '../../pathModule';
import kebabCaseModule from 'kebab-case-module';
import snakeCaseModule from 'snake_case_module';
import weirdMixedSeparatorModule from 'weird_mixed-separatorModule';
import dotSeparatedModule from 'dot.separated.module';
import scopedPackage from '@elg/scoped-package';
import  from '.';
"
`;

exports[`buildImportLines should handle destructure-only module import 1`] = `
"import { one, two } from 'a';
import { two } from 'b';
"
`;

exports[`buildImportLines should handle empty module import 1`] = `
"import 'a';
import 'b';
"
`;

exports[`buildImportLines should merge and deduplicate multiple import maps 1`] = `
"import a from 'a';
import b, { two } from 'b';
import { one, zero } from 'c';
import { minusOne } from './../../moreCharacters';
import 'd';
import 'e';
"
`;
